File: apps/frontend/src/lib/cleanup-disk-mode.ts
import type { Settings } from '../types';

export type CleanupDiskModeId = 'auto' | 'hdd' | 'standard';

export function normalizeCleanupDiskMode(raw: unknown): CleanupDiskModeId {
  const s = String(raw ?? 'auto').toLowerCase();
  if (s === 'hdd' || s === 'sequential') return 'hdd';
  if (s === 'standard' || s === 'ssd' || s === 'fast') return 'standard';
  return 'auto';
}

export function cleanupDiskModeLabel(mode: CleanupDiskModeId): string {
  switch (mode) {
    case 'hdd':
      return 'HDD / sequential';
    case 'standard':
      return 'Standard (follow scan workers)';
    default:
      return 'Auto (conservative on large batches)';
  }
}

export function cleanupDiskModeDescription(mode: CleanupDiskModeId): string {
  switch (mode) {
    case 'hdd':
      return 'Deletes one folder at a time. Best for mechanical drives and thousands of node_modules.';
    case 'standard':
      return 'Uses the same parallel worker count as Scan behavior → Performance (up to 6–8 trees at once).';
    default:
      return 'Caps parallelism on large batches and honors Low scan workers. Recommended default.';
  }
}

export function effectiveCleanupDiskMode(settings: Settings | null): CleanupDiskModeId {
  return normalizeCleanupDiskMode(settings?.cleanup_disk_mode);
}

/** Hint when a huge delete might benefit from HDD mode. */
export function suggestHddModeForSelection(
  selectedCount: number,
  mode: CleanupDiskModeId,
): boolean {
  return selectedCount > 400 && mode === 'auto';
}

File: apps/desktop/src-tauri/src/engine/cleanup_disk_mode.rs
//! Delete parallelism profile for mechanical vs fast disks (v0.6.6).

use super::delete_parallel::delete_parallelism_from_scan_mode;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CleanupDiskMode {
    /// Respect scan workers with batch caps (existing v0.6.5 behavior).
    Auto,
    /// One tree at a time — best for HDDs and huge batches.
    Hdd,
    /// Always use scan performance preset caps (may parallelize).
    Standard,
}

pub fn parse_cleanup_disk_mode(raw: &str) -> CleanupDiskMode {
    match raw.trim().to_ascii_lowercase().as_str() {
        "hdd" | "sequential" => CleanupDiskMode::Hdd,
        "standard" | "ssd" | "fast" => CleanupDiskMode::Standard,
        _ => CleanupDiskMode::Auto,
    }
}

/// Effective parallel tree deletes for in-place cleanup.
pub fn delete_parallelism_for_cleanup(
    disk_mode: CleanupDiskMode,
    scan_concurrency_mode: &str,
    item_count: usize,
) -> usize {
    match disk_mode {
        CleanupDiskMode::Hdd => 1,
        CleanupDiskMode::Standard => {
            delete_parallelism_from_scan_mode(scan_concurrency_mode, item_count)
        }
        CleanupDiskMode::Auto => {
            let mode = if scan_concurrency_mode == "low" {
                "low"
            } else {
                scan_concurrency_mode
            };
            delete_parallelism_from_scan_mode(mode, item_count)
        }
    }
}

pub fn cleanup_disk_mode_label(mode: CleanupDiskMode) -> &'static str {
    match mode {
        CleanupDiskMode::Auto => "auto",
        CleanupDiskMode::Hdd => "hdd",
        CleanupDiskMode::Standard => "standard",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hdd_mode_is_always_sequential() {
        assert_eq!(
            delete_parallelism_for_cleanup(CleanupDiskMode::Hdd, "high", 10_000),
            1
        );
    }

    #[test]
    fn auto_respects_low_scan_mode() {
        assert_eq!(
            delete_parallelism_for_cleanup(CleanupDiskMode::Auto, "low", 50),
            2
        );
    }

    #[test]
    fn parse_aliases() {
        assert_eq!(parse_cleanup_disk_mode("sequential"), CleanupDiskMode::Hdd);
        assert_eq!(parse_cleanup_disk_mode("standard"), CleanupDiskMode::Standard);
    }
}
